src/stream/xml_reader.rs

```diff
@@ -1,221 +1,262 @@
+    }
+
+    fn read_content(&mut self, buffer: &mut Vec<u8>) -> Result<String, Error> {
         loop {
-            match self.xml_reader.next() {
-                Ok(XmlEvent::Characters(s)) => return Ok(s),
-                Ok(event @ XmlEvent::EndElement { .. }) => {
-                    self.queued_event = Some(event);
-                    return Ok("".to_owned());
-                }
-                Ok(XmlEvent::EndDocument) => {
-                    return Err(self.with_pos(ErrorKind::UnclosedXmlElement))
+            match self.read_xml_event(buffer)? {
+                XmlEvent::Text(text) => {
+                    let pos = self.xml_reader_pos();
+                    let unespcaped = text
+                        .unescaped()
+                        .map_err(|err| ErrorKind::from(err).with_position(pos))?;
+                    return String::from_utf8(unespcaped.to_vec())
+                        .map_err(|_| ErrorKind::InvalidUtf8String.with_position(pos));
                 }
-                Ok(XmlEvent::StartElement { .. }) => {
-                    return Err(self.with_pos(ErrorKind::UnexpectedXmlOpeningTag));
+                XmlEvent::End(element) => {
+                    self.closed_element = Some(element.local_name().into());
+                    return Ok("".to_owned());
                 }
-                Ok(XmlEvent::ProcessingInstruction { .. }) => (),
-                Ok(XmlEvent::StartDocument { .. })
-                | Ok(XmlEvent::CData(_))
-                | Ok(XmlEvent::Comment(_))
-                | Ok(XmlEvent::Whitespace(_)) => {
-                    unreachable!("parser does not output CData, Comment or Whitespace events");
+                XmlEvent::Eof => return Err(self.with_pos(ErrorKind::UnclosedXmlElement)),
+                XmlEvent::Start(_) => return Err(self.with_pos(ErrorKind::UnexpectedXmlOpeningTag)),
+                XmlEvent::PI(_)
+                | XmlEvent::Empty(_)
+                | XmlEvent::Comment(_)
+                | XmlEvent::CData(_)
+                | XmlEvent::Decl(_)
+                | XmlEvent::DocType(_) => {
+                    // skip
                 }
-                Err(err) => return Err(from_xml_error(err)),
             }
         }
     }
 
-    fn next_event(&mut self) -> Result<XmlEvent, XmlReaderError> {
-        if let Some(event) = self.queued_event.take() {
-            Ok(event)
-        } else {
-            self.xml_reader.next()
-        }
-    }
-
-    fn read_next(&mut self) -> Result<Option<OwnedEvent>, Error> {
+    fn read_next(&mut self, buffer: &mut Vec<u8>) -> Result<Option<OwnedEvent>, Error> {
         loop {
-            match self.next_event() {
-                Ok(XmlEvent::StartDocument { .. }) => {}
-                Ok(XmlEvent::StartElement { name, .. }) => {
+            if let Some(closed_name) = self.closed_element.take() {
+                // Check the corrent element is being closed
+                match self.element_stack.pop() {
```

Comment:
[`Reader::check_end_names`](https://docs.rs/quick-xml/latest/quick_xml/reader/struct.Reader.html#method.check_end_names) looks like it might do this check for us.

Reply:
oh, and it's set to `true` by default, so we can skip this check completely. Good catch!

Reply:
I think that means we can get rid of everything related to `closed_element` and `element_stack`, which would reduce the code complexity quite a bit :)

Reply:
Yep, it does mean that, I've implemented it now.